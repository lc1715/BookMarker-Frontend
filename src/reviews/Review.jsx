import { useState, useContext } from 'react';
import UserContext from '../auth/UserContext';
import BookMarkerApi from "../api/api";
import EditReviewForm from './EditReviewForm';
//Material UI
import Box from '@mui/material/Box';
import Typography from '@mui/material/Typography';
import Modal from '@mui/material/Modal';
import IconButton from '@mui/material/IconButton';
import Divider from '@mui/material/Divider';
// import { useTheme } from '@mui/material/styles';
// import useMediaQuery from '@mui/material/useMediaQuery';


//shows review. has option to edit or delete review
function Review({ userReview, setReviewChange, setRanFirstScroll, setScrollToReview }) {
    const [editReviewForm, setEditReviewForm] = useState(false);
    const { currentUser } = useContext(UserContext);


    async function openEditReviewForm() {
        setEditReviewForm(true);
    };

    function closeEditReviewForm() {
        setEditReviewForm(false);
    }

    async function deleteReview() {
        try {
            await BookMarkerApi.deleteReview(userReview.id, currentUser.username);

            setReviewChange(true);
            setScrollToReview(false);
            setRanFirstScroll(false);
        } catch (err) {
            console.log(err);
        }
    };

    //styles the modal for edit review form
    const style = {
        position: 'absolute',
        top: '50%',
        left: '50%',
        transform: 'translate(-50%, -50%)',
        width: 620,
        bgcolor: 'background.paper',
        border: '1px solid #000',
        boxShadow: 24,
        py: 3,
        width: {
            xs: 348, sm: 446, md: 600, lg: 620
        }
    };

    return (
        <div>
            <Box sx={{ ml: 3, mr: { xs: 2, lg: 3 } }}>
                <h2 style={{ marginTop: '30px', marginBottom: '30px', textDecoration: 'underline' }}>
                    My Review:
                </h2>

                <Typography sx={{ fontSize: 19, wordBreak: "break-word" }}>{userReview.comment}</Typography>

                <Box sx={{ mb: 1, mt: 1, display: 'flex', justifyContent: 'space-between', alignItems: 'center' }}>
                    <p style={{ fontSize: '17px', color: '#5A5A5A' }}>
                        {userReview.username} - {new Date(userReview.created_at).toLocaleDateString()}
                    </p>

                    <Box sx={{ display: 'flex', gap: 1 }}>
                        <IconButton onClick={openEditReviewForm} color='success' aria-label="edit review"  >
                            <i class="fa-solid fa-pen-to-square"></i>
                        </IconButton>
                        <IconButton onClick={deleteReview} color='secondary' aria-label="delete review">
                            <i class="fa-solid fa-trash"></i>
                        </IconButton>
                    </Box>
                </Box>

                <Modal
                    open={editReviewForm}
                    onClose={closeEditReviewForm}
                    aria-labelledby="modal-modal-reviewForm"
                    aria-describedby="modal-modal-reviewForm"
                >
                    <Box sx={style}>
                        <EditReviewForm userReview={userReview} setReviewChange={setReviewChange} setEditReviewForm={setEditReviewForm} />
                    </Box>
                </Modal>
            </Box>

            <Divider sx={{ borderWidth: 1, mb: 1 }} />
        </div >
    );
};

export default Review;
