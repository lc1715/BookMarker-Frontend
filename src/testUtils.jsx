import UserContext from "./auth/UserContext";

const demoUser = {
    id: 1,
    username: 'testUser',
    email: 'test@email.com',
    volume_ids: ['e3', 't5'],
}

const UserProvider =
    ({ children, currentUser = demoUser, hasSavedBook = () => (true) }) => (
        <UserContext.Provider value={{ currentUser, hasSavedBook }}>
            {children}
        </UserContext.Provider>
    );

export { UserProvider };
