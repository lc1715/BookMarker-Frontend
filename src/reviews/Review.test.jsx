import { it, expect } from 'vitest';
import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { UserProvider } from "../testUtils";
import Review from './Review';

const review = {
    id: 1,
    comment: 'user review',
    created_at: 'date1',
    user_id: 2,
    username: 'user1',
    volume_id: 'vol1'
}

it('renders the Review component', () => {
    render(
        <MemoryRouter>
            <UserProvider>
                <Review userReview={review} />
            </UserProvider>
        </MemoryRouter>
    )
})

it('matches with snapshot', async () => {
    const { asFragment } = render(
        <MemoryRouter>
            <UserProvider>
                <Review userReview={review} />
            </UserProvider>
        </MemoryRouter>
    )

    expect(asFragment()).toMatchSnapshot();
})

it('has correct review data', async () => {
    render(
        <MemoryRouter>
            <UserProvider>
                <Review userReview={review} />
            </UserProvider>
        </MemoryRouter>
    )

    expect(screen.getByText('My Review:')).toBeInTheDocument();
    expect(screen.getByText('user review')).toBeInTheDocument();
});