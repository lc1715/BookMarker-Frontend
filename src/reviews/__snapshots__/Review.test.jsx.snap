// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches with snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="MuiBox-root css-wj6yh5"
    >
      <h2
        class="MuiBox-root css-2m1700"
      >
        My Review:
      </h2>
      <p
        class="MuiTypography-root MuiTypography-body1 css-1xy95es-MuiTypography-root"
      >
        user review
      </p>
      <div
        class="MuiBox-root css-144zmi7"
      >
        <p
          class="MuiBox-root css-6n9v02"
        >
          user1 - Invalid Date
        </p>
        <div
          class="MuiBox-root css-1i27l4i"
        >
          <button
            aria-label="edit review"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSuccess MuiIconButton-sizeMedium css-jp4vw-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1oyn1cm-MuiSvgIcon-root"
              data-testid="EditIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.996.996 0 0 0-1.41 0l-1.83 1.83 3.75 3.75z"
              />
            </svg>
          </button>
          <button
            aria-label="delete review"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSecondary MuiIconButton-sizeMedium css-1o4uo8s-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1oyn1cm-MuiSvgIcon-root"
              data-testid="DeleteIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth css-1kjxvo5-MuiDivider-root"
    />
  </div>
</DocumentFragment>
`;
